policy/common/predicateinfo.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+type PredicateInfo struct {
+	Name string
+	Type string
+
+	ContributorInfo *ContributorInfo
+	BranchInfo      *BranchInfo
+	FileInfo        *FileInfo
+	LabelInfo       *LabelInfo
+	CommitInfo      *CommitInfo
+	StatusInfo      *StatusInfo
+	TitleInfo       *TitleInfo
+}
+
+type ContributorInfo struct {
+	Organizations []string
+	Teams         []string
+	Users         []string
+	Author        string
+	Contributors  []string
+}
+
+type BranchInfo struct {
+	Patterns []string
+	Branch   string
+}
+
+type FileInfo struct {
+	Paths              []string
+	IgnorePaths        []string
+	ChangedFiles       []string
```

Comment:
It would be nice if there was a way to further generalize these structs to reduce the number of different types to manage. For example, these three fields of `FileInfo`, `LabelInfo`, `StatusInfo`, and `TitleInfo` all implement a similar concept:

```go
type PatternInfo struct {
  IncludePatterns []string
  ExcludePatterns []string
  ActualValues    []string
}
```

Some of those fields will be empty or only contain one value, but conceptually, all of these predicates have a list of required values/patterns, a (optional) list of values/patterns to ignore/exclude/fail, and then one or more actual values matched against the patterns.

Add in a `ValueName` field or something and when rendering, you can do:

```
Predicate {{.Name}} is satisfied because the {{.PatternInfo.ValueName}}:
{{range .PatternInfo.ActualValues}}* {{.}}
{{end}}
matched:
{{range .PatternInfo.IncludePatterns}}* {{.}}
{{end}}
```

which renders as:

```
Predicate OnlyChangedFiles is satisfied because the files:

* src/test/FooTest.java
* src/test/BarTest.java

matched:

* ^src/test/.*$
```

You might not even need the lists since there's usually a single `(pattern, value)` combination that passes or fails the rule.

Reply:
I am using the `PatternInfo` you suggested for labels, branches and title, but not for files. The files have an `IgnorePath` field which indicates the fields to ignore. This piece of information cannot be included in `PatternInfo` unless we add a new field specifically for it. Also, currently the way we check if a file matches a pattern is use a function called `anyMatches()`, this function returns a boolean. So if we want to use a single `(pattern, value)` combination, we will need to change that function.